test/testNodes.ts

```diff
@@ -0,0 +1,782 @@
+	interface ASTClass extends ASTNode {}
+	interface ASTBlock extends ASTNode {}
+}
+
+declare module '../src/server/ast/literals'
+{
+	interface ASTInt extends ASTNode {}
+	interface ASTFloat extends ASTNode {}
+	interface ASTStringLit extends ASTNode {}
+	interface ASTCharLit extends ASTNode {}
+	interface ASTBool extends ASTNode {}
+	interface ASTNull extends ASTNode {}
+}
+
+declare module '../src/server/ast/operations'
+{
+	interface ASTFunctionCall extends ASTNode {}
+	interface ASTUnaryOp extends ASTNode {}
+	interface ASTBinaryOp extends ASTNode {}
+	interface ASTPrefixOp extends ASTUnaryOp {}
+	interface ASTPostfixOp extends ASTUnaryOp {}
+	interface ASTDeref extends ASTUnaryOp {}
+	interface ASTInvert extends ASTUnaryOp {}
+	interface ASTMul extends ASTUnaryOp {}
+	interface ASTAdd extends ASTUnaryOp {}
+	interface ASTShift extends ASTUnaryOp {}
+	interface ASTBit extends ASTUnaryOp {}
+	interface ASTRel extends ASTUnaryOp {}
+	interface ASTBetween extends ASTNode {}
+	interface ASTLogic extends ASTUnaryOp {}
+	interface ASTAssign extends ASTNode {}
+}
+
+function eqOk()
+{
+	return Ok({
+		message: '',
+		pass: true,
+	})
+}
+
+function eqFail(message: string, path: TestPath)
+{
+	let errorLocation = ''
+	if (path.latestToken)
+	{
+		const line = path.latestToken.location.start.line
+		const column = path.latestToken.location.start.character
+		errorLocation = `At line: ${line + 1}, column: ${column}\n`
+	}
+
+	return Err({
+		message: `${errorLocation}${message} in ${path.path}`,
+		pass: false,
+	})
+}
+
+type NextT = {toString: () => string}
+function nextPath<T extends NextT>(path: TestPath, expected: T): TestPath
+{
+	const newPath = {...path}
+	newPath.path = `${path.path} -> ${expected.toString()}`
+	return newPath
+}
+
+function matchTypeAndToken(received: ASTNode, expected: ASTNode, path: TestPath): EqualPass
+{
+	path.latestToken = received.token
+
+	if (received.type !== expected.type)
+	{
+		const msg = `Expected ASTType: '${ASTType[expected.type]}', received '${ASTType[received.type]}'`
+		return eqFail(msg, path)
+	}
+
+	received.token.isEqualNode(expected.token, path).unwrap()
+	return eqOk()
+}
+
+type EqualType<T> = {isEqualNode: (expected: T, path: TestPath) => EqualPass} & NextT
+function undefinedEqual<T extends EqualType<T>>(
+	received: T | undefined,
+	expected: T | undefined,
+	valName: string, path: TestPath,
+): EqualPass
+{
+	if (received === undefined && expected)
+		return eqFail(`Value '${valName}' expected to be defined. Was undefined`, path)
+	if (received && expected === undefined)
+		return eqFail(`Value '${valName}' expected to be undefined. Was defined`, path)
+
+	if (expected && received)
+	{
+		const next = nextPath(path, expected)
+		received.isEqualNode(expected, next).unwrap()
+	}
+
+	// Both are undefined so they are equal
+	return eqOk()
+}
+
+function undefinedListEqual<T extends EqualType<T>>(
+	received: (T | undefined)[],
+	expected: (T | undefined)[],
+	valName: string, path: TestPath,
```

Comment:
`path: TestPath,` on its own line please for consistency.

We'd take using line flow semantics too - but this half-way house between
```ts
function undefinedListEqual<T extends EqualType<T>>(received: (T | undefined)[],
	expected: (T | undefined)[], valName: string, path: TestPath) : EqualPass
```
and
```ts
function undefinedListEqual<T extends EqualType<T>>(
	received: (T | undefined)[],
	expected: (T | undefined)[],
	valName: string,
	path: TestPath
): EqualPass
```
is the problem (you can apply this to our nitpick on `undefinedEqual()` above too)